tools/gcs-usage-recommender/python/backfill.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2019 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import json
+import sys
+import logging
+from typing import Dict, List
+from google.cloud import resource_manager
+from google.cloud import storage
+
+
+def get_storage_client() -> storage.Client:
+    """Retrieves GCS client.
+
+    Returns:
+        storage.Client object
+    """
+    return storage.Client()
+
+
+def get_resource_manager_client() -> resource_manager.Client:
+    """Retrieves Resource Manager Cleint
+
+    Returns:
+        resource_manager.Client object
+    """
+    return resource_manager.Client()
+
+
+def get_project_ids(resource_mgr_client: resource_manager.Client,
+                    organization_id: int) -> List[str]:
+    """Generates list of project_ids within a GCP org.
+
+    Args:
+        resource_mgr_client: resource_manager.Client
+        organization_id: Integer representing GCP org ID
+
+    Returns:
+        List of strings holding all project IDs.
+    """
+    project_id_list = []
+    project_filter = {
+        'parent.id': organization_id,
+        'parent.type': 'organization'
+    }
+    for project in resource_mgr_client.list_projects(project_filter):
+        project_id_list.append(project.project_id)
+    return project_id_list
+
+
+def get_buckets(project_ids: List[str],
+                gcs_client: storage.Client) -> List[Dict[str, str]]:
+    """Retrieves list of metadata for all buckets in a GCP org.
+
+    Args:
+        project_ids: List of strings holding project IDs
+        gcs_client: storage.Client object
+
+    Returns:
+        List of dictionaries mapping bucket-level metadata.
+    """
+    output_list = []
+    try:
+        for project_id in project_ids:
```

Comment:
In a large organization with 1000s of projects and potentially millions of buckets this could be slow have we run any benchmarks at scale?
 
Seems like this could greatly benefit from by parallelizing the `project_ids` loop. However, this might be premature optimization if we think this is truly a onetime activity where we could tolerate some latency.

Reply:
I ran it on our org, as well as a test GCS org from an SA. Both of which performed surprisingly well. Also, since it is a onetime activity, the small amount of latency that there is can probably be tolerated.

Reply:
Agreed.
Out of curiosity, do you have rough number of projects / buckets and time your benchmark took? Is this in line with the project / bucket cardinality of our larger customers?